### ISSUE TITLE
log4j issue with gatk3.8 and picard2.18

### ISSUE THREAD

**Issue by alexadn** [ngs-fzb/MTBseq_source#77]
Hi, 
we have been using MTBseq for some years but due to the log4j-issues end of 2021 and consequential IT resitrictions on our systems we had to put it out of service, as it uses older versions of gatk and picard that are affected by the security leak.

From what I saw on  https://gatk.broadinstitute.org/hc/en-us/articles/4412955012379-Important-GATK-version-update-required-to-address-Log4j-security-vulnerability and https://github.com/broadinstitute/picard/issues/1763 all gatk versions <4.2.4.0 and picard <2.26.9 are affected. 
 
Does this problem affects MTBseq at all using only some functionalities of gatk3.8 and picard (at the moment v2.18.29 comes with the conda installation)? Or is there any way to use the most recent versions of gatk/picard with mtbseq? 
I was not able to implement it so far. We would really appreciate to use MTBseq again.
Thanks a lot!!

**Comment by cutpatel** [comment]
Unfortunately there is no way to use a higher gatk version, as functionality has changed. Using a higher picard version is up to you and would work. You could just install a newer version.

The only other option you have is to skip the refinement step, which usually not makes much or any difference. We were already discussing to also skip this completely.

For now you would need to modify the mtbseq script:
You would need to remove the check for the tools starting in line 134 for gatk and picard. 

You need to remove the complete step starting in line 409 until 446.

From line 454 you have to point to the Bam directory and the bam files. You need to exchange every code that mentions gatk. You can take the code lines from the TBrefine step (409 until 446) which point to the Bam directory and files.

Cheers

**Comment by alexadn** [comment]
Thank you for the recommendations. Good to hear that picard is working in with newer versions and that the refinement step can be skipped in this situation.

I did the suggested changes in the MTBseq script. I tried to go without the refinement steps and and to point to the bam files for everything afterwards. However, TBpile is running without error but also without any output files. Then TBlist crashes as there are no files in Mpileup
Alternatively, I just did teh changes in the script before TBrefine. I then ran TBbwa only, copied the *bam\*bai from Bam to GATK_Bam as *gatk.bam/*gatk.bai and tried to go on with TBpile, but ended up the same.

If I am not wrong I think TBpile is also using gatk (line 476 of MTBseq and TBpile.pm)?
Is there any way to skip also that and go on with files fitting in the subsequent steps?

Thank you again!!!

**Comment by cutpatel** [comment]
You are right there are more changes to do. GATK is actually not used in the TBpile, but we already prepared something to make use of it, which we never did in th end.

You need to change every "gatk" from 450-487 to "bam" and you need to remove 481 the $GATK_dir completely

In addition you need to remove in the lib/TBpile.pm line 22. Remove the ".gatk\" in line 33 and need also  to change every  other "gatk" to a "bam". mainly GATK_OUT to BAM_OUT.

For your alternativ to move the bam files to the GATK directory and to rename them only the removals in line 481 and in 22 in TBpile.pm should be neccessary.

More changes could be possible, but maybe you can try?

**Comment by alexadn** [comment]
Thanks again for the help. 

The changes in line 22 of TBpile did the trick. 

However, as I am not that fit in Perl and had a faster successful outcome I stayed with the alternative just skipping TB refine while copying and renaming the Bams to GATK_Bams. 

With the more comprehensive solution I did some round but always found another path or variable which had to be renamed. 

The only thing is that until now, running "...TBfull .... --continue" still stops after TBlist saying " All position lists have been already created..." 
I guess this has something to do with counting steps but skipping one. But from that point it can simply be restarted without any error with TBvariants to be finished normally. I will check if I can get rid of that unnecessary stop in between if I have time. But anyway, the pipeline is running functionally again.

Thank you!!
